### ISSUE TITLE
Performances of Argon4 browser app

### ISSUE THREAD

**Issue by that-ben** [argonjs/argon#29]
First of all, thanks to people working on this project, it's very nice to see a peek preview of the technology standards that the next version Safari might finally be able to comply with... and if not, then Argon beautifully fills this gap!

OK now, about the performances of Argon4 browser app on iOS (might be the same on other platforms, did not check yet) is there a reason why it's slower by a factor of 200% to 250% ? I mean without the AR stuff, just the WebGL stuff alone.  Safari to Argon4, same page, same WebGL features, nothing AR related at the moment.  On Safari I get for example 34-40fps, on Argon4 the same exact page gets 11-16fps...

Any reason for this and could we improve?

Thanks.

**Comment by speigg** [comment]
Do you have a link to a page that you are using to measure the performance? Argon4 apps run in the same WKWebView with the same JIT javascript runtime optimizations as the Mobile Safari browser.

**Comment by that-ben** [comment]
It's my own app and I am not ready to publicly share the URL for it just yet, but try any of the more complex WebGL demos and you'll see the performance problems, for instance the CLOTH animation from ThreeJS.org here: http://threejs.org/examples/#webgl_animation_cloth

(Click the ANIMATION/CLOTH link to launch the demo)

That runs at 35-36fps with some random peaks to 40fps on iPhone 5 iOS 9.3 Safari.  Now on Argon4 browser, this runs at 28-29fps which is not too bad of a drop, because this is a very basic demo, but once you begin having more physics, textures, vertex count, etc... it becomes quite obvious.  My app that is 200-250% (!) slower on Argon4 also makes use of WebSockets and sends out DATA 3 times a second continuously, maybe Argon4 is also less efficient at I/O? Altough this demo does not communicate at all and still proves to be slower by 30% than the original iOS Safari.

**Comment by blairmacintyre** [comment]
(Sorry, didn't mean to close that)

I suspect the issue is that the video may still be running in the background.  The act of reading the camera and rendering the video is non-trivial, and (because we wanted to focus on getting the app out) didn't worry immediately about managing it as aggressively as we might.  so I think it may still be running whenever a custom reality is not in use.

Gheric, is this still the case?

**Comment by that-ben** [comment]
Interesting, I read earlier an older post on here about performance & energy conservation, I didn't know if the current release I got from the App Store this morning included those tweaks or not yet.  I understand 99% of the web pages loaded from Argon4 will have AR integrated in them, but if we could at least have the camera feed shut down when it's not in use, it would be much better for energy and performance indeed.  Let me know, I'll be your new official annoying tester if you want :)

**Comment by blairmacintyre** [comment]
I agree, it is going to be important to figure this out.  

At the very least, we should be able to turn off the camera when the front page is not showing an AR view.

One question is figuring out the right way to ask developers what capabilities they are using.  For example, if an app is only using Vuforia for tracking, we could shut down both CoreMotion (including the compass) and the GPS, which would be a great savings.   But, do developers understand that they don't need full device orientation if they are using Vuforia, since vision tracking makes it appear that they need to know the device orientation.

It may be as simple as saying "I need the geopose (position + orientation)" or not, and we turn it off if nobody is using it.  But, you could imagine separating them "I need geolocation" "I need device orientation" ... however, do do full orientation we need location because that's needed to correct "North" based on location.  

It'll happen, though.

**Comment by speigg** [comment]
Yes, the video reality is still rendering in the background when a regular webpage is displayed. We can (will) fix that. But I'm not sure that this is the main reason for the performance drop. We currently use the highest accuracy setting for geolocation, and that's a huge energy drain, and my guess is it also uses a considerable ammount of CPU, though I haven't profiled it yet. So, there may be a lot to gain in terms of performance (CPU usage and battery conservation) by configuring the geolocation service to report with much lower accuracy.

**Comment by blairmacintyre** [comment]
I think the video is the problem.  You can see it when you run an argon web app in mobile safari and it's overlaid on the white background:  the app is much faster.  There is non-trivial overhead in the video capture/renderer.

I think the higher accuracy GPS consumes battery;  I can't imagine in consumes CPU, since it's not doing the work on the CPU.

But, yes, profiling will show what's what!

**Comment by speigg** [comment]
As I understand, the issue with GPS is not that processing the GPS signal is computationally expensive, but that it keeps the CPU from sleeping.

**Comment by blairmacintyre** [comment]
That would be surprising to me, but I'm willing to believe anything.  I could imagine, for example, that this is also about doing "as good a job as possible" fusing the GPS and the wifi info to give as good a location estimate as possible, which might imply continually processing the wifi signal strength.  But even then, I would imagine this is a regular, but small, amount of processing.

But I still have a hard time imagining this has anything like the cost of the video (in CPU and battery)

**Comment by blairmacintyre** [comment]
Regardless:  we need to find ways to intelligently turn off 
- video
- gps
- orientation/compass
when they are not needed
